groovy/gradle/src/org/netbeans/modules/gradle/api/NbGradleProject.java

```diff
@@ -0,0 +1,371 @@
+    public Quality getAimedQuality() {
+        return project.getAimedQuality();
+    }
+
+    public boolean isUnloadable() {
+        return getQuality().worseThan(Quality.SIMPLE);
+    }
+
+    public Preferences getPreferences(boolean shared) {
+        Preferences ret = shared ? sharedPrefs : privatePrefs;
+        if (ret == null) {
+            if (shared) {
+                ret = sharedPrefs = ProjectUtils.getPreferences(project, NbGradleProject.class, true);
+            } else {
+                ret = privatePrefs = ProjectUtils.getPreferences(project, NbGradleProject.class, false);
+            }
+        }
+        return ret;
+    }
+
+    private void fireProjectReload() {
+        project.fireProjectReload(false);
+    }
+
+    private void doFireReload() {
+        detachResourceWatchers();
+        support.firePropertyChange(PROP_PROJECT_INFO, null, null);
+        attachResourceWatchers();
+    }
+
+    private void detachResourceWatchers() {
+        for (File resource : resources) {
+            try {
+                FileUtil.removeFileChangeListener(FCHSL, resource);
+            } catch (IllegalArgumentException ex) {
+                assert false : "Something is wrong with the resource handling";
+            }
+        }
+        resources.clear();
+    }
+
+    private void attachResourceWatchers() {
+        //Never listen on resource changes when only FALLBACK quality is needed
+        if (project.getAimedQuality() == Quality.FALLBACK) return;
+
+        Collection<? extends WatchedResourceProvider> all
+                = project.getLookup().lookupAll(WatchedResourceProvider.class);
+        for (WatchedResourceProvider pvd : all) {
+            resources.addAll(pvd.getWatchedResources());
+        }
+        for (File resource : resources) {
+            try {
+                FileUtil.addFileChangeListener(FCHSL, resource);
+            } catch (IllegalArgumentException ex) {
+                assert false : "Something is wrong with the resource handling";
+            }
+        }
+    }
+
+    public void addPropertyChangeListener(PropertyChangeListener propertyChangeListener) {
+        support.addPropertyChangeListener(propertyChangeListener);
+    }
+
+    public void removePropertyChangeListener(PropertyChangeListener propertyChangeListener) {
+        support.removePropertyChangeListener(propertyChangeListener);
+    }
+
+    /**
+     * Retrieves the watcher for the given project. Usually the project watcher
+     * can be retrieved from the project Lookup. This implementation does not
+     * use the project Lookup, so it can be used inside of the constructors
+     * of ProjectServiceProvider implementations.
+     *
+     * @param project the project to query
+     * @return the watcher of the project or {@code null} if the given project
+     *         is not a Gradle project.
+     */
+    public static NbGradleProject get(Project project) {
+        if (project instanceof NbGradleProjectImpl) {
```

Comment:
I was expecting following hierarchy:
```java
NbGradleProject implements Project;
NbGradleProject internally delegates to NbGradleProjectImpl;
```
but haven't checked the implications this would have on the implementation.